### PR TITLE
Migrate over GEM notice support

### PR DESCRIPTION
**Pull request by EvanHerman**
@jonathanbardo As discussed this is migrated over from the system plugin branch.

~New functions and files have been introduced, the version has been bumped to `1.1.4` and the readme.txt has been updated accordingly~.

**Update:**
An informational notice is displayed back to the user on the initial dashboard page load, and then `wpem_gem_notice` option is deleted from the database to prevent subsequent page loads.

**Note:** Needs translations.

_______________
### Original PR:

Display an admin notice with information about the GEM plugin on fresh WordPress installs.

This is a persistent notice, with the ability to dismiss. Once the 'x' has been clicked, the notice will no longer appear back to the user.

Admin notice

The .pot file needs to be updated with the new strings.

### CODE REVIEW

File: includes/settings.php [hunk 1]
```diff
@@ -545,15 +545,6 @@ public function display_settings_page() {
 
 				<?php if ( ! empty( $forms->signups ) ) : ?>
 					<div id="forms" class="panel">
-						<h3><?php esc_html_e( 'Reach Your Fans', 'godaddy-email-marketing' ); ?></h3>
```
Comment (jonathanbardo) [inline includes/settings.php]: @EvanHerman Why did you remove that?

File: godaddy-email-marketing.php [hunk 2]
```diff
@@ -71,11 +71,13 @@ public static function instance() {
...
-		add_action( 'widgets_init',   array( $this, 'register_widget' ) );
-		add_action( 'init',           array( $this, 'register_shortcode' ), 20 );
-		add_action( 'admin_notices',  array( $this, 'action_admin_notices' ) );
+		add_action( 'plugins_loaded',                  array( $this, 'i18n' ) );
+		add_action( 'init',                            array( $this, 'init' ) );
+		add_action( 'widgets_init',                    array( $this, 'register_widget' ) );
+		add_action( 'init',                            array( $this, 'register_shortcode' ), 20 );
+		add_Action( 'admin_enqueue_scripts',           array( $this, 'register_admin_scripts' ) );
+		add_action( 'admin_notices',                   array( $this, 'action_admin_notices' ) );
+		add_action( 'wp_ajax_dismiss_gem_notice',      array( $this, 'delete_wpem_gem_notice_option' ) );
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman We can reduce the spacing between the filter name and the array. Only 1 space between the comman of the biggest filter name.

File: godaddy-email-marketing.php [hunk 3]
```diff
@@ -71,11 +71,13 @@ public static function instance() {
...
-		add_action( 'plugins_loaded', array( $this, 'i18n' ) );
-		add_action( 'init',           array( $this, 'init' ) );
-		add_action( 'widgets_init',   array( $this, 'register_widget' ) );
-		add_action( 'init',           array( $this, 'register_shortcode' ), 20 );
-		add_action( 'admin_notices',  array( $this, 'action_admin_notices' ) );
+		add_action( 'plugins_loaded',                  array( $this, 'i18n' ) );
+		add_action( 'init',                            array( $this, 'init' ) );
+		add_action( 'widgets_init',                    array( $this, 'register_widget' ) );
+		add_action( 'init',                            array( $this, 'register_shortcode' ), 20 );
+		add_Action( 'admin_enqueue_scripts',           array( $this, 'register_admin_scripts' ) );
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman I think there is a typo in `add_Action`.

File: godaddy-email-marketing.php [hunk 4]
```diff
@@ -231,9 +233,50 @@ public function deactivate() {
 	}
 
 	/**
+	 * Enqueue admin scripts
+	 *
+	 * @since 1.1.4
+	 */
+	public function register_admin_scripts() {
+
+		if ( get_option( 'wpem-gem-notice' ) ) {
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman Might be better to use an early return statement here.

```php
if ( ! get_option( 'wpem-gem-notice' ) ) {

    return;

}
```

File: godaddy-email-marketing.php [hunk 5]
```diff
@@ -231,9 +233,50 @@ public function deactivate() {
...
+		}
+
+	}
+
+	/**
 	 * Displays the admin notice.
 	 */
 	public function action_admin_notices() {
+
+		if ( get_option( 'wpem-gem-notice' ) ) {
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman Let's add a private function to get rid of this `if`. `$this->display_gem_notice` or something that makes sense.

File: godaddy-email-marketing.php [hunk 6]
```diff
@@ -256,6 +299,21 @@ public function action_admin_notices() {
 			<?php
 		}
 	}
+
+	/**
+	 * Delete the wpem-gem-notice option from the database
+	 *
+	 * @return boolean
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman This function doesn't return a boolean.

File: godaddy-email-marketing.php [hunk 7]
```diff
@@ -231,9 +233,50 @@ public function deactivate() {
...
+		if ( get_option( 'wpem-gem-notice' ) ) {
+
+			?>
+
+			<div class="gem-notice notice notice-success is-dismissible">
+				<p>
+					<?php
+						printf(
+							__( 'Your website has a superpower: Email marketing. %1$s.', 'gd-system-plugin' ),
+							'<a href="' . admin_url( 'options-general.php?page=gem-settings' ) . '">' . __( 'Learn More', 'gd-system-plugin' ) . '</a>'
```
Comment (jonathanbardo) [inline godaddy-email-marketing.php]: @EvanHerman The text domain is not the right one.

File: js/gem-notice.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+/**
```
Comment (jonathanbardo) [inline js/gem-notice.js]: @fjarrett I think we do since .is-dismissible only adds the X and the animation and nothing else.

Comment (jonathanbardo) [inline js/gem-notice.js]: We need to fire the ajax somehow so the alert doesn't come back

File: includes/settings.php [hunk 9]
```diff
@@ -545,15 +545,6 @@ public function display_settings_page() {
...
-						<p><?php
-							printf(
-								esc_html__( 'Email marketing makes it easier than ever to turn casual visits into lasting relationship. You\'re already collecting subscribers, now you just need to start emailing them. It only takes a few moments to %1$screate an email marketing campaign%2$s.', 'godaddy-email-marketing' ),
-								'<a href="https://gem.godaddy.com" target="_blank">',
-								'</a>'
-							);
-							?>
-						</p>
```
Comment (EvanHerman) [inline includes/settings.php]: @fjarrett No, I think I originally forked this from another branch that didn't yet have this block of code.

File: js/gem-notice.js (hunk 8 above)
Comment (EvanHerman) [inline js/gem-notice.js]: The dismiss function works out of the box, but it only hides the container. The option will still exist in the database causing the notice to display again.

This file fires an AJAX function that removes the option from the database, preventing the notice from displaying on subsequent page loads.

Comment (fjarrett) [inline js/gem-notice.js]: I don't like that we have to do so much for this. Seems like it should be simpler...

Comment (fjarrett) [inline js/gem-notice.js]: @jonathanbardo @EvanHerman why not just delete the `wpem_gem_notice_option` on `shutdown` if we know it exists? Then the message would fire only once the first time you go to the Admin, and go away with a click of the X or reload of any page.

Comment (EvanHerman) [inline js/gem-notice.js]: @fjarrett We can go that route, but I was worried that the notice would get overlooked after the first page load. It is less code that route. If you think that would be best I can adjust the PR accordingly.

File: godaddy-email-marketing.php [hunk 10]
```diff
@@ -151,6 +158,7 @@ public function init() {
 
 		// Initialize settings.
 		if ( is_admin() ) {
+			$this->display_gem_notice = get_option( 'wpem_gem_notice' );
```
Comment (fjarrett) [inline godaddy-email-marketing.php]: @EvanHerman Seems like we should be able to look for this inside the `action_admin_notices` method and don't need this private property at all. Or did you already try that?

Comment (EvanHerman) [inline godaddy-email-marketing.php]: @fjarrett You're right. This was originally placed in a private variable because it was used in more than one location (DRY). Without the AJAX this is un-necessary. Removing it now.

File: godaddy-email-marketing.php [hunk 11]
```diff
@@ -234,6 +234,26 @@ public function deactivate() {
...
+						printf(
+							__( 'Your website has a superpower: Email marketing. %1$s.', 'godaddy-email-marketing' ),
+							'<a href="' . admin_url( 'options-general.php?page=gem-settings' ) . '">' . __( 'Learn More', 'godaddy-email-marketing' ) . '</a>'
+						);
+					?>
+				</p>
+			</div>
+
+			<?php
+
```
Comment (fjarrett) [inline godaddy-email-marketing.php]: @EvanHerman I say we just delete the option right here after displaying the notice.
